src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory_QueryLambdaRewriter.vb

```diff
@@ -0,0 +1,93 @@
+                Dim expression As BoundExpression = Nothing
+                If Not _rangeVariableMap.TryGetValue(node.RangeVariable, expression) Then
+                    ' _rangeVariableMap should contain an entry for the range variable, except for error cases. 
+                    Debug.Assert(node.HasErrors OrElse node.RangeVariable.Type.IsErrorType())
+                    Return node
+                End If
+
+#If DEBUG Then
+                ' Range variable reference should be rewritten to a parameter reference, or a call or a property access.
+                ' We clone these bound nodes in QueryLambdaRewriterPass2 to avoid dag in the generated bound tree.
+                ' If the LocalRewriter is changed to generate more kind of bound nodes for range variables, we should handle these in QueryLambdaRewriterPass2.
+                ' Below assert helps us to stay in sync with the LocalRewriter.
+                Select Case expression.Kind
+                    Case BoundKind.Parameter
+                    Case BoundKind.Call
+                    Case BoundKind.PropertyAccess
+                        Exit Select
+                    Case Else
+                        Debug.Fail($"Unexpected bound kind '{expression.Kind}' generated for range variable rewrite by method '{NameOf(LocalRewriter.PopulateRangeVariableMapForQueryLambdaRewrite)}'")
+                End Select
+#End If
+
+                Return expression
+            End Function
+        End Class
+
+        Private NotInheritable Class QueryLambdaRewriterPass2
+            Inherits BoundTreeRewriterWithStackGuard
+            Private ReadOnly _uniqueNodes As HashSet(Of BoundExpression)
+
+            Public Sub New()
+                _uniqueNodes = New HashSet(Of BoundExpression)
+            End Sub
+
+            Private Function HandleNode(Of T As BoundExpression)(node As T, cloneNode As Func(Of T, T)) As T
+                If Not _uniqueNodes.Add(node) Then
+                    node = cloneNode(node)
+                    _uniqueNodes.Add(node)
+                End If
+
+                Return node
+            End Function
+
+            Public Overrides Function VisitParameter(node As BoundParameter) As BoundNode
+                node = DirectCast(MyBase.VisitParameter(node), BoundParameter)
+                Return HandleNode(node, cloneNode:=Function(n) New BoundParameter(node.Syntax, node.ParameterSymbol, node.IsLValue, node.SuppressVirtualCalls, node.Type, node.HasErrors))
+            End Function
+
+            Public Overrides Function VisitCall(node As BoundCall) As BoundNode
+                node = DirectCast(MyBase.VisitCall(node), BoundCall)
+                Return HandleNode(node, cloneNode:=Function(n) New BoundCall(node.Syntax, node.Method, node.MethodGroupOpt, node.ReceiverOpt, node.Arguments, node.ConstantValueOpt, node.IsLValue, node.SuppressObjectClone, node.Type, node.HasErrors))
```

Comment:
>Return HandleNode(node, cloneNode:=Function(n) New BoundCall(node.Syntax, node.Method, node.MethodGroupOpt, node.ReceiverOpt, node.Arguments, node.ConstantValueOpt, node.IsLValue, node.SuppressObjectClone, node.Type, node.HasErrors)) [](start = 16, length = 233)

This is probably unnecessary, because it will all start with parameter access, which will be cloned, forcing base implementation to clone the call.